src/components/BranchMenu.tsx

```diff
@@ -0,0 +1,315 @@
+   * @returns React component
+   */
+  constructor(props: IBranchMenuProps) {
+    super(props);
+
+    const repo = this.props.model.pathRepository;
+    this._addListeners();
+
+    this.state = {
+      filter: '',
+      branchDialog: false,
+      current: repo ? this.props.model.currentBranch.name : '',
+      branches: repo ? this.props.model.branches : []
+    };
+  }
+
+  /**
+   * Callback invoked when a component will no longer be mounted.
+   */
+  componentWillUnmount() {
+    this._removeListeners();
+  }
+
+  /**
+   * Renders the component.
+   *
+   * @returns fragment
+   */
+  render() {
+    return (
+      <div className={wrapperClass}>
+        <div className={filterWrapperClass}>
+          <div className={filterClass}>
+            <input
+              className={filterInputClass}
+              type="text"
+              onChange={this._onFilterChange}
+              value={this.state.filter}
+              placeholder="Filter"
+              title="Filter branch menu"
+            />
+            {this.state.filter ? (
+              <button className={filterClearClass}>
+                <ClearIcon
+                  titleAccess="Clear the current filter"
+                  fontSize="small"
+                  onClick={this._resetFilter}
+                />
+              </button>
+            ) : null}
+          </div>
+          <input
+            className={newBranchButtonClass}
+            type="button"
+            title="Create a new branch"
+            value="New Branch"
+            onClick={this._onNewBranchClick}
+          />
+        </div>
+        <div className={listWrapperClass}>
+          <List disablePadding>{this._renderItems()}</List>
+        </div>
+        <NewBranchDialog
+          open={this.state.branchDialog}
+          model={this.props.model}
+          onClose={this._onNewBranchDialogClose}
+        />
+      </div>
+    );
+  }
+
+  /**
+   * Renders menu items.
+   *
+   * @returns fragment array
+   */
+  private _renderItems() {
+    return this.state.branches.map(this._renderItem, this);
+  }
+
+  /**
+   * Renders a menu item.
+   *
+   * @param branch - branch
+   * @param idx - item index
+   * @returns fragment
+   */
+  private _renderItem(branch: Git.IBranch, idx: number) {
+    // Perform a "simple" filter... (TODO: consider implementing fuzzy filtering)
+    if (this.state.filter && !branch.name.includes(this.state.filter)) {
+      return null;
+    }
+    return (
+      <ListItem
+        button
+        title={`Switch to branch: ${branch.name}`}
+        className={classes(
+          listItemClass,
+          branch.name === this.state.current ? activeListItemClass : null
+        )}
+        key={idx}
+        onClick={this._onBranchClickFactory(branch.name)}
+      >
+        <span className={listItemIconClass} />
+        {branch.name}
+      </ListItem>
+    );
+  }
+
+  /**
+   * Adds model listeners.
+   */
+  private _addListeners(): void {
```

Comment:
@fcollonval Sorry, but I don't follow. I included comments explaining why the branch menu listens for each of the three events.

Listening for a `head` change is not enough because, if I create a new branch from the current branch and do so outside of JupyterLab, the `HEAD` does not change, and, thus, the branch change does not get reflected in the UI. The signal which does get emitted when the branch changes, but the `HEAD` does not change, is the `status` signal.

More of an edge case, but a repository change does not necessarily entail a `HEAD` change either. If I had two clones of the same repository sitting side-by-side, I could switch repositories, but still point to the same hash.

Yes, this does mean that multiple event listeners could be triggered, potentially even all 3; however, per the implementation proposed in this PR, this should not trigger a re-render unless model state changes in the interim. In which case, the perf cost is likely to be negligible, and I think acceptable in order to account for edge cases, particularly when mixing Git management inside and outside of JupyterLab (which is something I do regularly).

IMO, I don't think we should have a separate poll for refreshing a list of branches, as multiple polls at different intervals leads to synchronicity issues. And further, even if we polled `refreshBranch`, I don't think that that method should emit a `headChanged` signal, particularly if the `HEAD` has not changed.

Lastly, this PR intentionally eschews the mixed data flow model, as mentioned in the OP and as was used previously, where child components rely on both unidirectional data flow (via properties) **and** querying the underlying extension model (e.g., using getters). Such a mixed model leads to synchronicity bugs, as encountered during development.

Instead, the implementation proposed in this PR relies exclusively on the model (and not parent components) as the source of truth. Moving toward a poll method would most likely entail reverting back to a mixed data flow (e.g., see the `HistorySideBar` component) where, presuming `refreshBranch` does not emit a `HEAD` change event, we'd have to set a `branches` property on the branch menu component to ensure that the UI does not drift from the underlying Git repository.

Furthermore, again presuming that we'd not move toward relying on polling and emitting `HEAD` change events, the fact that changing a repository currently invokes `refresh` (which invokes `refreshBranch`) is only relevant to the implementation currently proposed by this PR insofar as the model's `_branches` property is updated, which the branch menu subsequently accesses after receiving one (or more) signal(s) that the list of branches has potentially changed.

In short, if potentially redundant signals is an overarching concern, I'd be open to throttling, but, in general, I am not particularly inclined to move toward polling or removing listeners which would potentially leave the UI in a state which deviates from the underlying Git repository.

Reply:
One further addendum: looking back through the `model` implementation, maybe what you meant is that `refresh` and thus a `status` change signal is emitted upon changing the repository. In which case, you're right. I will remove the `repositoryChanged` listeners.